### PR TITLE
Enable GM station(s) from clients

### LINKED ISSUE #543: Remote GM
How hard would it be to program a remote GM console for headless servers?

### PR DESCRIPTION
**Pull request by amir-arad**
enable Gm station from clients
- tweaks and AI infographics does not work on non-server GM stations
fixes #543

### CODE REVIEW

File: src/GMActions.cpp [hunk 1]
```diff
@@ -45,6 +48,71 @@ void GameMasterActions::onReceiveClientCommand(int32_t client_id, sf::Packet& pa
...
+            int index;
+            packet >> index;
+            int selectedItemsLeft;
+            packet >> selectedItemsLeft;
+            PVector<SpaceObject> selection;
+            while (selectedItemsLeft--) {
+                int selectedItemId;
+                packet >> selectedItemId;
+                P<SpaceObject> object;
+                if (game_server)
```
Comment (daid) [inline src/GMActions.cpp]: this should always be running on the server, so the if (game_server) is alway true

File: src/gameGlobalInfo.h [hunk 2]
```diff
@@ -76,6 +76,8 @@ class GameGlobalInfo : public MultiplayerObject, public Updatable
 
     //List of script functions that can be called from the GM interface (Server only!)
     std::list<GMScriptCallback> gm_callback_functions;
+    //List of names of gm_callback_functions scripts (replicated to clients)
```
Comment (daid) [inline src/gameGlobalInfo.h]: I understand why this list is here, but now we have duplicate data, name is in both std::vectors. Would be cleaner to remove the name from the GMScriptCallback (which might not need to exist then and can be just a ScriptSimpleCallback object?)

Comment (JordanLongstaff) [reply src/gameGlobalInfo.h]: http://www.cplusplus.com/reference/algorithm/remove_if/

Comment (amir-arad) [reply src/gameGlobalInfo.h]: nevermind, I just RTFM-ed.

**Comment by JordanLongstaff** [comment]
I got build errors with this. I think you need to declare `class CpuShip;` in GMActions.h.

Also, you might consider changing CMD_SEND_COMM_TO_PLAYER_SHIP to 0x000A (hex for 10).

**Comment by JordanLongstaff** [comment]
Actually, no, you're just missing the imports of cpuShip.h and wormHole.h.

**Comment by JordanLongstaff** [comment]
Oh, no...I guess you actually do have those imports. Weird that I didn't see them.

Anyway, I tried clicking the "Start Server" button and nothing happened, so I clicked it again and got an assertion error in gameGlobalInfo.cpp at line 10.

**Comment by JordanLongstaff** [comment]
Turns out that I was having some kind of network issue, actually. For whatever reason, EmptyEpsilon wasn't able to open port 35666.

**Comment by amir-arad** [comment]
> Actually, no, you're just missing the imports of cpuShip.h and wormHole.h.

@JordanLongstaff I went over it a few times now and I dont see anything missing. in `GMActions.h` there is no use of CpuShip or Wormhole so no need to declare or import. and in `GMActions.cpp` both `.h` files are included. See [here](https://github.com/daid/EmptyEpsilon/blob/1043a6d1ee41d58dd4ab7e3f8cf259997006a94a/src/GMActions.cpp#L5-L6)

**Comment by amir-arad** [comment]
any chances of this getting merged?

**Comment by amir-arad** [comment]
Hi @daid I heard you might have missed some github notifications so i'm taking the chance of bugging you.
I'm lazily working on other stuff that is based on this PR (lower level GM control over cpu ships). it would be simpler for me to know if this PR is/isn't gonna make it in. I respect specific rejections or fix requests, as well as principal "feature not worth the complexity".

**Comment by amir-arad** [comment]
ping

**Comment by daid** [comment]
Oops? :-)

Code wise, looks quite fine. Only thing I'm wondering about is how well the GM to ship chat works with multiple GMs connected. As that code is already fragile with just a single GM screen...

**Comment by amir-arad** [comment]
@daid could you elaborate or suggest scenarios?
I did some sanity tests with 2 GM stations and 1 player ship but it was a while back. seemed to be working smoothly except for the UX for the GMs.
I was thinking about a switch for selectively enabling chats with specific ships in each GM station, so that any ship only talks with one GM at any given time. alternatively a GM comms station that only does chats. but that was for internal GM crew management purposes. I would love to hear more about the engineering concerns.

**Comment by daid** [comment]
Just a concern, nothing concrete jumps to mind, just something that I wondered how well you checked.

**Comment by amir-arad** [comment]
I didn't test it in "production" yet. I run 0-2 games a year, and this year was dry so far :(

**Comment by k9lego** [comment]
@amir-arad Daid mentioned on the USN that he's concerned about security issues relating to this. I could test this in a production game on the 15th if you could get this secure.

**Comment by daid** [comment]
By security I mean, it allows any client to do GM actions. And it allows and client to execute lua code on the server.

This could be fixable by making the GM client password protected, and setting a flag on the PlayerInfo object as soon as the correct password was entered, and checking this flag before and GM related function is executed on the server.

**Comment by amir-arad** [comment]
I've used this feature in a 40 hours larp so I can say it works in production. 

My upcoming game is a few weeks from now so i won't be available anytime soon to add the security feature for the sake of merging.

**Comment by k9lego** [comment]
Sounds good!

**Comment by Kilted-Klingon** [comment]
I would just like to add my 2 cents here in this thread -- I also would find it very useful to have the option to choose a GM station from a client in order to spread the GM responsibilities among a team.  I have tested one of @amir-arad versions, and it seems to work fine.  However, I would definitely want to have some sort of security password enabled through the scenario script.

**Comment by Kilted-Klingon** [comment]
@amir-arad , @daid , and @k9lego , I volunteer to help test capabilities.  Now that I can build both Linux and Windows versions from the GitHub repositories, I certainly don't mind helping out.

**Comment by amir-arad** [comment]
work continues on #965

**Pull request closed without merging**